File: fleet-management-system-backend/src/main/java/com/damocles/fleet/fleetmanagementsystembackend/service/TrailerService.java
package com.damocles.fleet.fleetmanagementsystembackend.service;
import com.damocles.fleet.fleetmanagementsystembackend.domain.TransportStatus;
import com.damocles.fleet.fleetmanagementsystembackend.domain.Trailer;
import com.damocles.fleet.fleetmanagementsystembackend.domain.TrailerStatus;
import com.damocles.fleet.fleetmanagementsystembackend.dto.trailer.CreateTrailerRequest;
import com.damocles.fleet.fleetmanagementsystembackend.dto.trailer.TrailerDTO;
import com.damocles.fleet.fleetmanagementsystembackend.exception.BusinessValidationException;
import com.damocles.fleet.fleetmanagementsystembackend.exception.NotFoundException;
import com.damocles.fleet.fleetmanagementsystembackend.exception.TrailerNotFoundException;
import com.damocles.fleet.fleetmanagementsystembackend.mapper.ITrailerMapper;
import com.damocles.fleet.fleetmanagementsystembackend.repository.ITrailerRepository;
import com.damocles.fleet.fleetmanagementsystembackend.repository.ITransportRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional
public class TrailerService implements ITrailerService {

    private final ITrailerRepository trailerRepository;
    private final ITrailerMapper trailerMapper;
    private final ITransportRepository transportRepository;

    @Override
    // Returns all trailers with assignment flags.
    public List<TrailerDTO> getAllTrailers() {
        return trailerRepository.findAll()
                .stream()
                .map(this::enrichTrailer)
                .toList();
    }

    @Override
    // Searches trailers by free-text query with paging.
    public Page<TrailerDTO> searchTrailers(String q, Pageable pageable) {
        String query = (q == null || q.isBlank()) ? null : q.trim().toLowerCase();

        if (query == null) {
            return trailerRepository.findAll(pageable).map(this::enrichTrailer);
        }

        return trailerRepository.search(query, pageable).map(this::enrichTrailer);
    }

    @Override
    // Fetches a trailer by id.
    public TrailerDTO getTrailerById(Long id) {
        Trailer trailer = trailerRepository.findById(id)
                .orElseThrow(() -> new TrailerNotFoundException(id));
        return enrichTrailer(trailer);
    }

    @Override
    // Creates a trailer record.
    public TrailerDTO createTrailer(CreateTrailerRequest req) {
        Trailer trailer = trailerMapper.toEntity(req);
        trailerRepository.save(trailer);
        return enrichTrailer(trailer);
    }

    @Override
    // Updates trailer fields by id.
    public TrailerDTO updateTrailer(Long id, CreateTrailerRequest req) {
        Trailer trailer = trailerRepository.findById(id)
                .orElseThrow(() -> new TrailerNotFoundException(id));
        trailerMapper.updateTrailerFromDto(req, trailer);
        return enrichTrailer(trailer);
    }

    @Override
    // Returns trailers that are currently available.
    public List<TrailerDTO> getAvailableTrailers() {
        return trailerRepository.findAvailable()
                .stream()
                .map(this::enrichTrailer)
                .toList();
    }

    @Override
    // Deletes a trailer only if it is not assigned to any transport.
    public void deleteTrailer(Long id) {
        Trailer trailer = trailerRepository.findById(id)
                .orElseThrow(() -> new TrailerNotFoundException(id));
        if (transportRepository.existsByTrailer_Id(id)) {
            throw new BusinessValidationException("Trailer is assigned to a transport and cannot be deleted");
        }
        trailerRepository.delete(trailer);
    }

    @Transactional
    // Updates trailer status when not assigned to active transport.
    public TrailerDTO updateStatus(Long id, TrailerStatus status) {
        Trailer t = trailerRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Trailer not found"));

        if (transportRepository.existsByTrailer_IdAndStatusIn(
                id,
                List.of(TransportStatus.ACCEPTED, TransportStatus.IN_PROGRESS)
        )) {
            throw new BusinessValidationException("Trailer is assigned to an active transport and status cannot be changed");
        }

        t.setTrailerStatus(status);
        return enrichTrailer(t);
    }

    private TrailerDTO enrichTrailer(Trailer t) {
        TrailerDTO base = trailerMapper.toDto(t);
        boolean assigned = transportRepository.existsByTrailer_Id(t.getId());
        boolean inProgress = transportRepository.existsByTrailer_IdAndStatusIn(
                t.getId(),
                List.of(TransportStatus.ACCEPTED, TransportStatus.IN_PROGRESS)
        );
        return new TrailerDTO(
                base.id(),
                base.name(),
                base.licensePlate(),
                base.payload(),
                base.volume(),
                base.trailerStatus(),
                assigned,
                inProgress
        );
    }
}

File: fleet-management-system-frontend/src/app/core/services/trailer/trailer.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map, Observable } from 'rxjs';
import { TrailerStatus } from '../../models/trailer-status.model';
import { CreateTrailerRequest, TrailerDTO } from "../../models/trailer.model";
import { Page } from "../../models/page.model";
import { buildHttpParams } from '../http-params.util';

export interface TrailerSearchParams {
  q?: string;
  page?: number;
  size?: number;
  sort?: string[];
}

@Injectable({ providedIn: 'root' })
export class TrailerService {
  constructor(private http: HttpClient) {}
  private readonly lookupSize = 200;

  // Fetches a limited list of trailers for lookups.
  getAll(size = this.lookupSize): Observable<TrailerDTO[]> {
    return this.list({ page: 0, size }).pipe(map(page => page.content));
  }

  // Fetches a trailer by id.
  getOne(id: number): Observable<TrailerDTO> {
    return this.http.get<TrailerDTO>(`/api/trailers/${id}`);
  }

  // Lists trailers with search and pagination.
  list(params: TrailerSearchParams): Observable<Page<TrailerDTO>> {
    const httpParams = buildHttpParams({
      q: params.q,
      page: params.page,
      size: params.size,
      sort: params.sort,
    });

    return this.http.get<Page<TrailerDTO>>('/api/trailers', { params: httpParams });
  }

  // Creates a trailer.
  createTrailer(req: CreateTrailerRequest): Observable<TrailerDTO> {
    return this.http.post<TrailerDTO>('/api/trailers', req);
  }

  // Fetches trailers that are currently available.
  getAvailable() {
    return this.http.get<TrailerDTO[]>('/api/trailers/available');
  }

  // Updates trailer status.
  changeStatus(id: number, status: TrailerStatus): Observable<TrailerDTO> {
    return this.http.patch<TrailerDTO>(`/api/trailers/${id}/status`, { status });
  }

  // Updates trailer details.
  updateTrailer(id: number, payload: CreateTrailerRequest): Observable<TrailerDTO> {
    return this.http.put<TrailerDTO>(`/api/trailers/${id}`, payload);
  }

  // Deletes a trailer by id.
  deleteTrailer(id: number): Observable<void> {
    return this.http.delete<void>(`/api/trailers/${id}`);
  }
}







